File: src/main/java/org/noxet/noxetserver/NoxetServer.java
package org.noxet.noxetserver;

import org.bukkit.World;
import org.bukkit.WorldCreator;
import org.bukkit.WorldType;
import org.bukkit.generator.ChunkGenerator;
import org.bukkit.plugin.java.JavaPlugin;
import org.noxet.noxetserver.commands.CommandRegistration;
import org.noxet.noxetserver.messaging.Motd;
import org.noxet.noxetserver.realm.RealmManager;

import java.io.File;
import java.util.Random;

public final class NoxetServer extends JavaPlugin {

    private static NoxetServer plugin;

    public enum WorldFlag {
        NEUTRAL, OVERWORLD, NETHER, THE_END, FLAT, VOID
    }

    public static boolean shouldAllowWorldPreservation = true;

    public enum ServerWorld {
        HUB("hub", null, true, true, WorldFlag.NEUTRAL),
        SMP_SPAWN("smp_spawn", RealmManager.Realm.SMP, true, true, WorldFlag.NEUTRAL),
        SMP_WORLD("smp_world", RealmManager.Realm.SMP, false, false, WorldFlag.OVERWORLD),
        SMP_NETHER("smp_nether", RealmManager.Realm.SMP, false, false, WorldFlag.NETHER),
        SMP_END("smp_end", RealmManager.Realm.SMP, false, false, WorldFlag.THE_END),
        ANARCHY_WORLD("anarchy", RealmManager.Realm.ANARCHY, false, false, WorldFlag.OVERWORLD),
        ANARCHY_NETHER("anarchy_nether", RealmManager.Realm.ANARCHY, false, false, WorldFlag.NETHER),
        ANARCHY_END("anarchy_end", RealmManager.Realm.ANARCHY, false, false, WorldFlag.THE_END),

        CANVAS_WORLD("canvas", RealmManager.Realm.CANVAS, false, true, WorldFlag.VOID);

        private final String worldName;
        private final RealmManager.Realm realm;
        private final boolean preservedWorld, safeZone;
        private final WorldFlag flag;

        ServerWorld(String worldName, RealmManager.Realm realm, boolean preservedWorld, boolean safeZone, WorldFlag flag) {
            this.worldName = worldName;
            this.realm = realm;
            this.preservedWorld = preservedWorld;
            this.safeZone = safeZone;
            this.flag = flag;
        }

        private WorldCreator getWorldCreator() {
            WorldCreator worldCreator = new WorldCreator(worldName);

            World.Environment environment = World.Environment.NORMAL;

            switch(getWorldFlag()) {
                case NETHER:
                    environment = World.Environment.NETHER;
                    break;
                case THE_END:
                    environment = World.Environment.THE_END;
                    break;
                case FLAT:
                    worldCreator.type(WorldType.FLAT);
                    worldCreator.generateStructures(false);
                    break;
                case VOID:
                    worldCreator.generator(new ChunkGenerator() {
                        @Override
                        @SuppressWarnings({"NullableProblems", "deprecation"})
                        public ChunkData generateChunkData(World world, Random random, int x, int z, BiomeGrid biome) {
                            return createChunkData(world);
                        }
                    });
                    break;
            }

            worldCreator.environment(environment);

            return worldCreator;
        }

        public World getWorld() {
            return NoxetServer.getPlugin().getServer().createWorld(getWorldCreator());
        }

        public boolean isPreserved() {
            return preservedWorld && shouldAllowWorldPreservation;
        }

        public boolean isSafeZone() {
            return safeZone;
        }

        public WorldFlag getWorldFlag() {
            return flag;
        }

        public RealmManager.Realm getRealm() {
            return realm;
        }
    }

    @Override
    public void onEnable() {
        plugin = this;

        getServer().getPluginManager().registerEvents(new Events(), this);

        Motd.loadQuotes();

        logInfo("Loading commands...");

        CommandRegistration.registerCommands();

        logInfo("Noxet plugin is ready.");
    }

    @Override
    public void onDisable() {
        logInfo("Noxet plugin stopped.");
    }

    public static NoxetServer getPlugin() {
        return plugin;
    }

    public File getPluginDirectory() {
        File dir = getDataFolder();

        if(!dir.mkdir() && (!dir.exists() || !dir.isDirectory()))
            throw new RuntimeException("Cannot create plugin directory.");

        return dir;
    }
    
    public static void logInfo(String message) {
        getPlugin().getLogger().info(message);
    }

    public static void logWarning(String message) {
        getPlugin().getLogger().warning(message);
    }

    public static void logSevere(String message) {
        getPlugin().getLogger().severe(message);
    }

    public static boolean isWorldPreserved(World world) {
        for(ServerWorld serverWorld : ServerWorld.values())
            if(serverWorld.getWorld() == world)
                return serverWorld.isPreserved();
        return false;
    }

    public static boolean isWorldSafeZone(World world) {
        for(ServerWorld serverWorld : ServerWorld.values())
            if(serverWorld.getWorld() == world)
                return serverWorld.isSafeZone();
        return false;
    }
}

File: src/main/java/org/noxet/noxetserver/NoxetServer.scala
package org.noxet.noxetserver

import org.bukkit.World.Environment
import org.bukkit.generator.ChunkGenerator
import org.bukkit.generator.ChunkGenerator.{BiomeGrid, ChunkData}
import org.bukkit.plugin.java.JavaPlugin
import org.bukkit.{World, WorldCreator, WorldType}
import org.noxet.noxetserver.commands.CommandRegistration
import org.noxet.noxetserver.messaging.Motd
import org.noxet.noxetserver.realm.RealmManager.Realm

import java.io.File
import java.util.Random

class NoxetServer extends JavaPlugin:
    override def onEnable(): Unit =
        NoxetServer.plugin = Some(this)

        getServer.getPluginManager.registerEvents(Events(), this)

        Motd.loadQuotes()

        NoxetServer.logInfo("Loading commands...")
        CommandRegistration.registerCommands()

        NoxetServer.logInfo("Noxet plugin is ready.")

    override def onDisable(): Unit = NoxetServer.logInfo("Noxet plugin stopped.")

    def getPluginDirectory: File =
        val dir = getDataFolder

        if (!dir.mkdir && (!dir.exists || !dir.isDirectory))
            throw new RuntimeException("Cannot create plugin directory.")

        dir

object NoxetServer:
    enum WorldFlag:
        case Neutral, Overworld, Nether, TheEnd, Flat, _Void

    enum ServerWorld(val worldName: String, val realm: Option[Realm], val preservedWorld: Boolean, val safeZone: Boolean, val flag: WorldFlag):
        case Hub extends ServerWorld("hub", None, true, true, WorldFlag.Neutral)
        case SmpSpawn extends ServerWorld("smp_spawn", Some(Realm.SMP), true, true, WorldFlag.Neutral)
        case SmpWorld extends ServerWorld("smp_world", Some(Realm.SMP), false, false, WorldFlag.Overworld)
        case SmpNether extends ServerWorld("smp_nether", Some(Realm.SMP), false, false, WorldFlag.Nether)
        case SmpEnd extends ServerWorld("smp_end", Some(Realm.SMP), false, false, WorldFlag.TheEnd)
        case AnarchyWorld extends ServerWorld("anarchy", Some(Realm.ANARCHY), false, false, WorldFlag.Overworld)
        case AnarchyNether extends ServerWorld("anarchy_nether", Some(Realm.ANARCHY), false, false, WorldFlag.Nether)
        case AnarchyEnd extends ServerWorld("anarchy_end", Some(Realm.ANARCHY), false, false, WorldFlag.TheEnd)

        case CanvasWorld extends ServerWorld("canvas", Some(Realm.CANVAS), false, true, WorldFlag._Void)

        private def getWorldCreator: WorldCreator =
            val worldCreator = WorldCreator(worldName)

            var environment = World.Environment.NORMAL

            flag match
                case WorldFlag.Nether => environment = Environment.NETHER
                case WorldFlag.TheEnd => environment = Environment.THE_END
                case WorldFlag.Flat =>
                    worldCreator.`type`(WorldType.FLAT)
                    worldCreator.generateStructures(false)
                case _Void =>
                    worldCreator.generator(new ChunkGenerator:
                        override def generateChunkData(world: World, random: Random, x: Int, z: Int, biome: BiomeGrid): ChunkData = createChunkData(world)
                    )

            worldCreator.environment(environment)

        def getWorld: World = NoxetServer.getPlugin.getServer.createWorld(getWorldCreator)

    private var plugin: Option[NoxetServer] = None

    def getPlugin: NoxetServer = plugin.get

    def shouldAllowWorldPreservation: Boolean = true

    def logInfo(message: String): Unit = getPlugin.getLogger.info(message)

    def logWarning(message: String): Unit = getPlugin.getLogger.warning(message)

    def logSevere(message: String): Unit = getPlugin.getLogger.severe(message)

    def isWorldPreserved(world: World): Boolean =
        ServerWorld.values.find(_.getWorld == world).exists(_.preservedWorld)

    def isWorldSafeZone(world: World): Boolean =
        ServerWorld.values.find(_.getWorld == world).exists(_.safeZone)